File: app/models/qc.rb
# frozen_string_literal: true

# Traits for QCs.
class Qc < ApplicationRecord
  include Broadcastable
  include BelongsToPlayer
  include HealthLevels
  include Intimacies
  include MotePool
  include Willpower
  include RankedModel

  ranks :sorting, with_same: :player_id
  ranks :chronicle_sorting, with_same: :chronicle_id

  with_options inverse_of: :qc, dependent: :destroy do
    has_many :qc_charms, -> { order(:sorting) }
    has_many :qc_merits, -> { order(:sorting) }
  end

  with_options dependent: :destroy do
    has_many :qc_attacks, -> { order(:sorting) }, as: :qc_attackable, inverse_of: :qc_attackable
    has_many :spells,     -> { order(:sorting) }, as: :sorcerer,      inverse_of: :sorcerer
    has_many :poisons,    -> { order(:sorting) }, as: :poisonable,    inverse_of: :poisonable
    # has_many :combat_actors, -> { order(:sorting) }, as: :actor, inverse_of: :actor
  end

  # Essence above 5 is explicitly mentioned in the book
  validates :essence, one_thru_ten_stat: true

  validates :grapple, :grapple_control,
            :hardness,
            :onslaught,
            :resolve, :guile, :appearance,
            :join_battle, :evasion, :parry, :soak,
            :movement, :senses,
            :feats_of_strength, :strength,
            numericality: { greater_than_or_equal_to: 0 }

  validates :actions,   json: { schema: Schemas::QC_ACTION }
  validates :resources, json: { schema: Schemas::RESOURCE }

  def entity_type
    'qc'
  end
  alias entity_assoc entity_type
  alias type entity_type
end

File: app/javascript/lca/types/qc.ts
import { PlayerAsset } from './_lib'
import { WithSharedStats } from './shared'

export interface QcAction {
  action: string
  pool: number
}

export interface QcStats {
  join_battle: number
  movement: number
  soak: number
  hardness: number
  appearance: number
  resolve: number
  guile: number
  evasion: number
  parry: number
  senses: number
  armor_name: string
}

export interface QcMerit {
  name: string
  latent: boolean
  magical: boolean
  body: string
  ref: string
}

export interface QcAttack {
  name: string
  pool: number
  damage: number
  overwhelming: number
  range: string
  tags: Array<string>
  qc_attackable_type: 'Qc' | 'Battlegroup'
  qc_attackable_id: number
}

export interface QC extends PlayerAsset, WithSharedStats, QcStats {
  name: string
  ref: string

  feats_of_strength: number
  shape_sorcery: number
  strength: number
  grapple: number
  grapple_control: number

  excellency: string
  actions: QcAction[]
  qc_attacks: number[]
  qc_charms: number[]
  qc_merits: number[]
  portrait_link: string
  type: 'qc'
}
